### DISCUSSION TITLE
Memory efficiency of reduction approach

### DISCUSSION THREAD

**Discussion by danbartl** [alan-turing-institute/sktime#1002] [General]
Hi all, thanks a lot for the great library, looking really forward to forecasting multivariate time series with `sktime`, especially with the imminent implementation of https://github.com/alan-turing-institute/sktime/pull/980.

@mloning @aiwalter @fkiraly 

However, I have a question regarding the current reduction approach via sliding window. The chosen approach of first creating the sliding windows in `make_reduction` and then applying transformers like `TsFreshFeatureExtractor` can become  memory inefficient when we have a small number of features generated, but a long window over which the features are generated to model long term dependence, which is often the case for multivariate time series. Take a look at the following code:

```python
from sklearn.ensemble import RandomForestRegressor

from sktime.datasets import load_airline
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose import make_reduction
from sktime.forecasting.model_selection import temporal_train_test_split
from sklearn.pipeline import make_pipeline
from sktime.transformations.panel.tsfresh import TSFreshFeatureExtractor
import pandas as pd
y = load_airline()
y_train, y_test = temporal_train_test_split(y)

regressor = make_pipeline(
    TSFreshFeatureExtractor(default_fc_parameters="minimal",show_warnings=False, disable_progressbar=True),
    RandomForestRegressor(),
)
forecaster = make_reduction(
    regressor, scitype="time-series-regressor", window_length=50
)
forecaster.fit(y_train)

fh = ForecastingHorizon(y_test.index, is_relative=False)
y_pred = forecaster.predict(fh)
```
`y` just has a `sys.getsizeof(y)` of 1744
In `make_reduction` in `_sliding_window_transform`, the `Zt` array that is created returns `sys.getsizeof(Zt) = 64592`.
The `Xt` object generated by `TsFreshFeatureExtractor` and that is actually needed has just a size of 4656.

Theoretically, `Xt` could have been directly generated from `y` without the need to create an all-encompassing array `Zt`.

While I understand that creating `Zt` is a very elegant approach that typically will not lead to issues for small univariate data, multivariate data can easily reach sizes of dozens of GB, so not increasing the required memory usage beyond what is absolutely necessary would probably help a lot of users. 

One possible way to fix that would be to move transformers that cover feature generation like `TSFreshFeatureExtractor` into the `make_reduction` step. In the end, `_sliding_window_transform`, is just the naive transform of returning the identity for every y, so allowing other transforms might be consistent with `sktime`.

Do you see these issues too, or am I missing something? 

Do you intend to make `sktime` suitable for such memory hungry uses cases and would you be open for such a change to move the location where transformers like `TSFreshFeatureExtractor` are applied, or do you see additional issues?

**Comment by fkiraly** [comment]
>Hi all, thanks a lot for the great library, 

Thank you!

>looking really forward to forecasting multivariate time series with sktime, especially with the imminent implementation of #980.

You're not creating any pressure at all here, are you 😄 

>However, I have a question regarding the current reduction approach via sliding window. The chosen approach of first creating the sliding windows in make_reduction and then applying transformers like TsFreshFeatureExtractor can become memory inefficient when we have a small number of features generated, but a long window over which the features are generated to model long term dependence, which is often the case for multivariate time series. 

Yes, I think you are right.

>One possible way to fix that would be to move transformers that cover feature generation like TSFreshFeatureExtractor into the make_reduction step. In the end, _sliding_window_transform, is just the naive transform of returning the identity for every y, so allowing other transforms might be consistent with sktime.

Hm, the problem is that you would have to do this for every transformer you could possibly compose the reducer with.
Already for `tsfresh`, due to the large size of features it creates, this seems like a herculean task?

Another problem that this creates is that we don't want `make_reduction` to be a god object that handles all possible reduction/transformer combinations, there are just too many of those.

One can, of course, adopt the paradigm of supporting just a small selection of combinations well, but then you completely lose the flexibility of the scikit-learn-like design, and are no longer a framework toolbox but an algorithm toolbox. Which has its place and justification, of course, but not what sktime is aiming at.

>Do you see these issues too, or am I missing something?
>Do you intend to make sktime suitable for such memory hungry uses cases and would you be open for such a change to move the location where transformers like TSFreshFeatureExtractor are applied, or do you see additional issues?

Generally, I think we're very aware of the issues of "memory hunger" and use of compute when doing compositions. We would very much like to solve this and solve this well.
Another example we have is time series forest, where we've noticed that there's a factor of 100-1000 between the composite implementation and the optimized "en bloc" implementation.

So it's worth doing this in the first case for the most frequently used combinations, but it's often not clear which combinations of components those are. And as we are entirely volunteer driven and unfunded, it might not be the first priority when compared to other important maintenance tasks.

The more high-tech way of doing this would be an optimization or compilation layer, that given a declarative specification creates an efficient computation graph under the hood, similar to what dask-ml does. But that's a whole project in itself...

Good ideas are very welcome - do you have any concrete suggestions that are not too time intensive, and/or which perhaps you could contribute, say, via PR?

**Reply by fkiraly** [reply]
(you could also join the dev days which are currently underway if interested in contributing)

**Reply by danbartl** [reply]
Thanks for the detailed and thorough explanation. I understand that it would be difficult/impossible to make an implementation for every transformer. But as you also observed it matters more for certain models, especially tree based ones, so maybe we can indeed narrow it to certain combinations. I am still learning more about this library, so maybe when I have a clearer picture of the library  -and the implementation of https://github.com/alan-turing-institute/sktime/pull/980  ;) :) - I can make a suggestion. 

In any case, thanks for clarifying your view, and looking forward to the next steps in this library.
